### ISSUE TITLE
Strengthen the definition of ``pure`` potentially disallowing e.g. ``codecopy`` - or "the future of ``pure``".

### ISSUE THREAD

**Issue by ekpyron** [ethereum/solidity#8153]
Came up in https://github.com/ethereum/solidity/issues/3835, in particular https://github.com/ethereum/solidity/issues/3835#issuecomment-574854014 states the issue.

The main question is: since ``view`` already allows ``staticcall``, is there any use for the coarse definition of ``pure``, that allows for expressions that are *not* compile-time constant?

Personally, I'd say we should officially strengthen pure, so that everything pure can be used in place of compile-time-constant expressions, e.g. the most basic definition of pure is "everything that can be used as array size of a statically sized array".

We *could* also move towards the opposite direction and e.g. *allow* reading the address of ``this`` in ``pure`` functions and weaken ``pure`` to mainly mean "does not read from storage". But then we need yet another stricter concept for compile-time-constant expressions. So I'm not a fan of that course of action :-).

**Comment by axic** [comment]
This is what pure was supposed to be, but it was softened up somehow :wink:

**Comment by chriseth** [comment]
I agree, pure functions should not be allowed to read from data that can be modified by a constructor.

**Comment by ekpyron** [comment]
I agree, but then ``SemanticInformation::invalidInPureFunctions`` should be extended by ``CODECOPY`` - and there may be others, we should check. I'll move this to "accepted in 0.7.0" right away then (unless we call it "bug" and want to do it earlier).

**Comment by MicahZoltu** [comment]
(sorry for the double-post, missed the comment that discussion about pure should be moved here)

@ekpyron Great breakdown (in the linked issue) of the two potential definitions of pure.  I have always treated pure as along the lines of "_if you call this method on this contract multiple times with the same inputs, you'll get the same result every time no matter what other conditions change between calls_".  This interpretation generally aligns with what I want as a dapp developer, and it also makes sense that `codecopy` would be allowed.  To me, `pure` does not mean "_this function would yield the same result no matter what contract it was implemented in_".

I can kind of see the merits to the other definitions, but IMO "it yields the same result if called multiple times" feels both the most useful and the most natural.

Can someone express the argument as to why a _user_ would want pure to mean "provides the same result no matter what bytecode the function is deployed with" is?  I want to be careful that we are implementing pure in the most pure (pun intended) way because it offers the best developer experience, not just because it is the strictest definition we can come up with.

**Comment by chriseth** [comment]
@MicahZoltu the only problem I see is a potential confusion that a pure method in the less restrictive sense yields a different result when moved to a different contract. Or put a different way: In the functional sense, pure means that there is no other state to take into account except the arguments to the function call.

On the other hand, when seen at the EVM level, then a pure function always reads the code, because that is what is executed, so maybe it should also be able to use codecopy.

Maybe we should ask the same question from the other end: How would you use "pure" as a feature when it is defined in the one or the other way?

**Comment by MicahZoltu** [comment]
Honestly, I don't use it much.  `view` gives _huge_ benefits over non-view, it is an even bigger step than from `payable` to `non-payable`.  `pure` however is a pretty small/marginal step.  It is useful sometimes during auditing, just as any compiler guarantees are, but it doesn't come up that often.

When it does come up in auditing for me, I am usually just using it as a shortcut to avoid going down some particular code path when I need to validate that a call won't read some dirty storage variable.  If `pure` didn't exist (only `view` did) I wouldn't really miss it that much personally, though I don't advocate for removing it as I'm a big fan of _any_ additional guarantees the compiler can make.

I would also support splitting `pure` into two different modifiers:
1. functionality can be determined by looking at the code of this function only
2. function will not read any storage variables or block details, but its execution may be dependent on static parts of the contract like the contract's code, the contract's hash, the contract's address, immutable variables, etc.

**Comment by chriseth** [comment]
Agreement from call: codecopy should not be pure (let alone for consistency with immutable).

@ekpyron which other things did you have in mind?

**Comment by ekpyron** [comment]
We should probably go through the list of opcodes to check if there's anything else, but ``codecopy`` was the main thing so far, that should be stricter.

**Comment by ekpyron** [comment]
This issue just came up again recently, when we realized that we - while we wanted to get all this sorted for 0.7.0 - don't really have consensus about ``pure`` and sorted this out for the upcoming 0.8.0 release :-). The plan now is to discuss this further and finally come to some consensus for 0.9.0.

While I think the two ways to think of ``pure`` have been described well already, let me offer another perspective on that that came to my mind:

I think that intuitively there is a difference between ``external pure`` and ``internal pure``, so there is no wonder, that there is disagreement about what a general ``public pure`` is supposed to be :-).
As an example ``msg.data`` came up recently. In an external function ``msg.data`` can effectively be considered the same as a function argument, so - thinking of ``pure`` as "a function only depends on inputs" - ``msg.data`` should be allowed in ``external pure`` functions.
Also, it does not make sense to think about "compile-time constantness" in ``external pure`` functions - e.g. I can never use the result of a call to an external function as static array size, ``function f() external pure returns(uint) {...} function g() public { uint[this.f()] x; }``.

However, ``internal pure`` functions are different. For those ``msg.data`` is *not* effectively the same as the function arguments, but rather "additional state". I can call the same internal pure function with the same arguments, expecting the same results, while having different contents in ``msg.data`` in the background.
Also, now for ``internal pure`` functions I would say "only depends on its inputs" effectively means it is "compile-time constant". So for internal pure functions, I would in fact expect to be able to use them in the size of statically sized arrays.

I'm definitely not saying that we should make the meaning of ``pure`` depend on visibility (especially since that would still not clarify what to do for ``public``), but maybe this helps in the discussion. I actually - contrary to my earlier position - might be starting to lean towards going for what I just described as ``external pure`` (having ``internal pure`` just be "usable in ``external pure``") and then have "compile-time constantness", which is quite a bit stricter, just be determined by the compiler via analysis (or there were suggestions to introduce something like ``constexpr`` for it, alternatively).

Still, even if we come to an agreement about this part, we still have to precisely clarify what counts as "state" and what as "input arguments" for those external pure functions, because the distinction is not always entirely clear.

**Comment by chriseth** [comment]
Your reasoning sounds consistent, @ekpyron . It might be not that confusing to strengthen the restrictions imposed by 'pure` at least for free functions.

**Comment by wjmelements** [comment]
Since the evaluation of a pure function depends on its immutable code, it seems silly to prevent codecopy or reading from immutable variables inside `pure` functions. Pushing a constant to the stack isn't conceptually different from pushing an immutable; both are in the code. But currently the constant allows `pure` but the immutable requires `view`.

**Comment by ekpyron** [comment]
The main reason for strengthening ``pure`` would be to be able to syntactically mark functions as compile-time-constant.
```
uint[f()] compileTimeConstantlySizedArray;
```
will only work sanely if ``f`` does *not* depend on its code and does not involve e.g. ``codecopy`` or especially immutable access (unless you want to run a recompilation in the constructor ;-)). It can very well involve pushing constants as much as it wants, though. So there *is* a conceptual difference between the two.

There is two ways to proceed with this:
1. ultimately decide to *weaken* pure to allow it to depend on code, thereby splitting "compile-time-constant" and ``pure`` (this would involve making ``immutable`` access pure) or
2. strengthen ``pure`` to really mean "compile-time-constant" (this would involve making ``codecopy`` ``view``).

Both are OK, but we have an inconsistent mixture right now. 1 would mean that we either need a different way to mark functions "compile-time constant" or merely rely on the compiler to determine whether it is or not (both would be feasible, though). The main question is whether 2 forces things to be ``view`` that would *really* benefit from being marked as a weaker ``pure`` (as long as I don't evaluate things at compile-time there is no difference between ``pure`` and ``view`` execution-wise at all, so it is mainly a question of which kind of annotation has more value).

**Comment by deluca-mike** [comment]
As someone who is heavy user (and proponent) of pure functions (and functional programming in general), I prefer option 1.

It seems reasonable to strive for objective technical accuracy, rather than striving for some subjective semantic accuracy.

Immutables end up as in-line litterals. The on-chain byte code for a function that returns a `constant`, `immutable`, or literal, is identical.

With a perfect decompiler that could translate any on-chain bytecode into human-readable Solidity (as one would write it), the following 4 smart contracts would have identically decompiled `foo` functions, all marked at `pure`:

```
contract ImmutableTest1 {
    function foo() external pure returns (uint256) { return uint256(4); }
}

contract ImmutableTest2 {
    uint256 internal constant _foo = uint256(4);
    function foo() external pure returns (uint256) { return _foo; }
}

contract ImmutableTest3 {
    uint256 internal immutable _foo;
    constructor() { _foo = uint256(4); }
    function foo() external view returns (uint256) { return _foo; }
}

contract ImmutableTest4 {
    uint256 internal immutable _foo;
    constructor(uint256 foo_) { _foo = foo_; }
    function foo() external view returns (uint256) { return _foo; }
}
```

In my opinion, `pure` should define what happens (and can't happen) on-chain, not what humans expect to happen before the contract is even created.

When you deploy a contract to the chain, you already need to wait for the receipt to be sure of its address (i.e. consider a factory that deploys children using the blockhash as a `create2` salt), so to interact with it, you need at least _some_ data/event from the chain. If you really needed to know what the resulting immutables were, if any, of a contract, just get more data from the chain (i.e. emit them in events, or expose externally).

**Comment by chriseth** [comment]
Thank you for your input, @deluca-mike ! I think your examples fall a bit short on the most generic case. Of course a function that uses an immutable that in fact is a compile-time constant can be pure in the most strict functional-programming sense.

The question is if code whose semantics depend on the time and circumstances it was deployed at can be considered pure or not.

**Comment by ekpyron** [comment]
Just because this has come up in the context of #12860 again:

I'm personally starting to change my mind on this issue. In https://github.com/ethereum/solidity/issues/3835#issuecomment-574854014 I basically argued that it makes a difference whether one considers ``external pure`` (does not read from storage nor from blockchain state) or ``internal pure`` (which one may interpret as "fully determined by the function arguments").

However, there is a different notion one can use for ``internal pure``, i.e. solely "it is admissible to call this internal function from an external pure function".

In this interpretation, it's perfectly fine to read arbitrary calldata and memory in ``internal pure`` functions - while arbitrary ``mload`` or ``calldataload`` (resp. access to ``msg.data``) clearly violates for a function to be fully determined by its arguments. Also the argument that ``returndatacopy`` is an issue in ``pure`` functions does not apply to the "admissible to call in an external pure function" interpretation.

I think the weak notion of it being admissible to call an internal function from an external pure function, is useful - for example in inheritance context (e.g. leaving an internal pure function unimplemented, but calling it, for a derived contract to provide an implementation, which may e.g. read ``msg.data``, but not state).

So I'm more and more leaning towards actually *weakening* ``pure`` to just the ``external pure`` interpretation.

There is still merit in decorating internal functions that are compile-time constant, i.e. that only depend on their immediate arguments and are pure in the classical functional sense. However, these days I'd rather introduce a new separate keyword for this property, like C++'s ``constexpr``.

**Comment by howardpen9** [comment]
follow from https://github.com/ethereum/solidity/issues/12829

**Comment by cameel** [comment]
I'm going to reopen this in connection with #16422, as I'd like to bring some consistency to the various pureness/constness levels before we go all in into comptime in Core Solidity.

I don't think there is any more need to strenghten "pure" all the way to a compile-time constant level now. We already de facto have that level - it's used for array sizes, storage layout base, arithmetic on rational literals. If we were to have a mutability modifier for it, I would not reuse `pure` for it and just introduce `comptime` as a new one. We also have another de facto level: runtime constant. Those are things you can assign to `constant` variables.

That frees up `pure` to be something broader. The question is how broad. I would generally agree with people above saying that it should cover deployed contract code. This would automatically fix issues like #15989. In my opinion `pure` would be more useful if it expressed the concept of a `view` function that, once deployed, will not give you a different result unless it explicitly gets different inputs via arguments. This means that we could allow accessing:
- contract's own deployed bytecode
- immutables
- address (and `this` in general)
- arguments stored in memory and calldata

On the other hand I would disallow:
- pure external calls
- Unrestricted calldata and memory reads (`msg.data`, `calldataload()`, `mload()`)

I would also propose extending `constant` variables to mean that rather than "runtime constant" to allow initializing constant variables with the result of any pure function. Or we could keep `constant` as is but introduce `pure` variables, though IMO there would be little need to ever use `constant` over `pure`.

---

As an alternative, in https://github.com/argotorg/solidity/issues/12260#issuecomment-3243050843 @k06a has a slightly different proposal. It matches mine regarding the things I would allow, but it's slightly broader, allowing anything that is "deterministic and side-effect-free". In practice this means that it does not disallow pure calls and read access to arbitrary memory and calldata locations via assembly.

Both are viable IMO, though one thing that I like less about it is that it would not make sense to align this with `constant`s, as a pure function under these conditions would no longer necessarily return a constant value.

**Comment by Czar102** [comment]
> In my opinion `pure` would be more useful if it expressed the concept of a `view` function that, once deployed, will not give you a different result unless it explicitly gets different inputs via arguments.

One thing to remember: even disallowing memory access via assembly in pure functions doesn't make the function deterministic. If the function is internally invoked by a non-pure one, then by violating the memory layout in the non-pure function (e.g. changing the free memory pointer to a lower value than some stored data or making a nonzero mstore at 0x60) may cause the pure function to behave differently than in case the memory layout wasn't violated, for example write to variables it wasn't supposed to. Luckily, I don't think that matters (see last paragraph).

There are also some nonintuitive actions that would violate the independence from other functions, like:
- `mstore()` can change the size of an existing array, making it possible to access values beyond it
- even assigning a value in assembly to a memory/calldata reference type will make its values (and hence external data) available
- declaring a struct without assigning a value will also create a pointer to zero which can be used to access arbitrary elements of memory after tricking Solidity to calculate any hash at scratch space `0x00` and then pretending a memory array has its length saved in that field

I don't we ever need `pure` functions to behave deterministically internally - as a function of explicit arguments declared in Solidity. Reasoning about them as comptime deterministic is only needed for the compiler, while runtime determinism of pure functions invoked by non-pure functions is not needed at all (or is it?). From all other perspectives, it's enough that `pure` functions are `view`, which is trivially true.
So, I think we can redefine determinism for `pure` functions: their execution may only depend on the calldata and is not limited to Solidity-declared arguments. External `pure` functions may invoke `pure` functions internally and these functions may access all data available, even if not explicitly passed in arguments. That doesn't change the determinism of the external `pure` function. Obviously, all `pure` functions may do anything they want aside from using forbidden opcodes/operations – without restricting assembly use.

tldr: don't look at function-level determinism, just at EVM-message-call determinism

**Comment by cameel** [comment]
So, if I understand correctly, you're circling back to the notion of "external pure" that was already posted above in https://github.com/argotorg/solidity/issues/8153#issuecomment-742447086? I guess this is still a valid option.

I focused more on the internal functions myself, because my interest in getting this in order came from constants (#16422). Being able to match them with `pure` sounds attractive to me, because it would reduce the number of concepts you have to be aware of. With "external pure" approach we're still left with `constant` as a distinct constness level between comptime and pure.

I'm not entirely sure if "external pure" would really be more useful to users, but I also have no strong arguments the other way. Comptime and `view` are two very concrete border conditions and anything we choose in between seems somewhat arbitrary.

In any case, some minor points about what you wrote:

> by violating the memory layout in the non-pure function (e.g. changing the free memory pointer to a lower value than some stored data or making a nonzero mstore at 0x60) may cause the pure function to behave differently than in case the memory layout wasn't violated

My idea was to actually disallow all memory writes as well as those are side-effects. That would include memory allocation. All the memory that is accessed would have to be passed in explicitly, via arguments.

Now, I admit that I did not fully think through all the consequences of that. It would probably be too restrictive. It would not allow you to have memory locals or temporaries, because you would not be able to allocate. Even initializing something with a zero pointer is really as an allocation in a way, so it would be off limits.

So I still think such pure functions would be deterministic in the sense we want. But you have a point in that maybe they would not be as useful as we want.

> declaring a struct without assigning a value will also create a pointer to zero which can be used to access arbitrary elements of memory after tricking Solidity to calculate any hash at scratch space `0x00`

A pointer to the zero slot (`0x60`), not a pointer to address zero. Scratch space does not overlap the zero slot, so you can't overwrite it by having the compiler write to the scratch space.

> Obviously, all `pure` functions may do anything they want aside from using forbidden opcodes/operations – without restricting assembly use.

What about pure external calls? Would you allow those? Or access to the code of external contracts (you can't prevent it anyway if you allow pure external calls)? These could still break EVM-call-level determinism when dealing with cases where contract code can change (e.g. using an address before and after a contract was deployed to it; or setting code via EIP-7701).

Also, what do you think about the notion of "internal pure" as something callable only from an external pure function (https://github.com/argotorg/solidity/issues/8153#issuecomment-1092895108)?

What about pure functions at creation time, when you can run code but the contract is not yet deployed?

**Comment by Czar102** [comment]
I don't think I properly explained my previous idea. I've refined it since and will show this simpler explanation (I may seem to contradict my previous comment, refer only to the following explanation).

### The goal
Let's remember the end goal of redefining `pure`. The aim is to be able to use `pure` functions in `constant` variable definition. This means that `pure` functions should be evaluable at compilation.
But, it DOESN'T mean they have to be fully deterministic. It means that it should be well-defined how will `pure` function's body and arguments be used to define the result at compilation time. For example, a nullary `internal pure` function returning `mload(0x40)` is not deterministic – it may return different values on invocation in different parts of the program. But, with clear evaluation rules, its result may be deterministically determined at compile time (I'll show that below).

It doesn't make sense to migrate the definition of `constant` from a compile-time constant because the next opportunity to avoid evaluation of logic is at the constructor. At that stage, `immutable` is the tool for exactly that, and it's way more expressive – there is no need to require determinism at that stage, and the idea of `constant` and `immutable` would converge.
Defining a function as `pure` must have its use case. Currently, (afaik) it simply means `view` with the additional guarantee that it doesn't read from the state. The latter is not useful anywhere, except when sanity-checking the properties of the function when auditing. On EVM interaction level, it means nothing. This shouldn't have justified adding a new keyword to the compiler; all new keywords should have actual uses aside from a nice compiler sanity-check that certain opcodes are not used.
(this is what I believe this issue is about)

### Basic properties
Just from the goal of compile-time determinism, we know that:
- invoking `pure external` functions in other contract must not be allowed, even if they are *declared* as pure (because the result is dependent on state: the code of the other contract)
- reading the code of itself or function pointers is not possible since they haven't been derived yet at compile-time; this clearly lies in the area of immutable functions
  - similarly, the address of the contract is unknown at compile-time
- reading immutables must be forbidden
- there is no concept of calldata at compile-time, so it doesn't make sense to read it
- `pure` precompiles are independent of a chain's state, so it's fine to allow them
- invoking `pure` functions with known implementations is fine (like `this.f(1337)`), even if they are `external` (although I do not recommend allowing this because it creates a wrong mental model and doesn't improve expressibility)

### Design
The design proposed has the goal of maximum expressibility in mind – any code that doesn't use forbidden opcodes (and doesn't revert and doesn't violate basic stack assumptions) creates a valid `pure` function.
With maximum expressibility, it comes obvious that it doesn't make sense for the compiler to have a known set of functions and just evaluate them to derive the result (like it currently is), because the functions are as expressible as the "pure EVM". So, the compiler should use an EVM interpreter (it can be greatly simplified from the "full EVM" interpreter, though).

Whenever an expression is to be evaluated by the compiler, it's done as follows:
1. the expression is compiled
2. a new EVM context is created; pointer `0x40` is initilized, memory-based arguments (like arrays) are saved in memory, function arguments are pushed on the initially empty stack; returndatasize is 0
3. the expression's code is run in a "pure EVM" interpreter; it obviously can't access state or context; some examples of forbidden opcodes:
  - `GAS`, `CHAINID`, `TIMESTAMP`
  - `SLOAD`, `SSTORE`, `TLOAD`, `TSTORE`
  - `CALLDATALOAD`, `CALLDATASIZE`
  - `CALL`
  - `STATICCALL` is not entirely forbidden – the contract called has to be a precompile (which is then executed), otherwise it is forbidden
4. if the execution fails, the compilation fails as well; if it succeeds, the result is left as the only element of the EVM stack

This design is well-defined (deterministic given deterministic EVM context setup), maximally expressive, and simultaneously most lightweight (the "pure EVM" interpreter should be much easier to implement than a full interpreter, especially given that there is only one active EVM context, only precompiles and basic opcodes have to be implemented, gas isn't measured, state never has to be read or written to).

### Final notes
Note that this design needs only `internal pure` functions, `external pure` functions would remain without a use case – pretty much equivalent to `external view` functions.

It will also be difficult to enforce purity for auditing while allowing `STATICCALL`, a higher-level analysis needs to be done to know that only a precompile can be invoked by some `STATICCALL` opcodes.

**Comment by cameel** [comment]
To ground the discussion a bit more, let me first clarify the terminology I'm using:
- *comptime* value: calculated during analysis - can affect type system (e.g. array sizes)
- *runtime* value: calculated by executing EVM code on chain
- *immutable* value: embedded in contract's code and read from there at runtime
- *constant* value: calculated once and guaranteed by the language not to modifiable

Some of these are mutually exclusive (comptime vs runtime obviously) but not all. For example you could say that our current `immutable` variables are *runtime immutable*. *comptime immutable*, on the other hand, would be what people often imagine `constant` to be - e.g. something that is calculated during compilation and stored in the code, while in reality they are *runtime constant*.

> The aim is to be able to use `pure` functions in `constant` variable definition. This means that `pure` functions should be evaluable at compilation.

Not really. It is possible to align the constness level of `constant` and `pure` regardless of which of these levels we settle on. runtime constant is just as viable as comptime. See https://github.com/argotorg/solidity/issues/8153#issuecomment-3800375483 above.

> Defining a function as `pure` must have its use case. Currently, (afaik) it simply means `view` with the additional guarantee that it doesn't read from the state. The latter is not useful anywhere, except when sanity-checking the properties of the function when auditing. On EVM interaction level, it means nothing. This shouldn't have justified adding a new keyword to the compiler; all new keywords should have actual uses aside from a nice compiler sanity-check that certain opcodes are not used.

Your view here is apparently that anything short of comptime is useless. I don't agree. Having constant in the language has many benefits and I don't think that the hard, EVM-level guarantees are the only ones worth taking into account:

1. **Interaction with the type system**. E.g. being able to use the value as array size, storage layout base or (in the future) as a parameter of generic types.
1. **Optimization**: pre-calculating the value of a constant expression usually means less gas and often smaller bytecode (though not always - e.g. it takes fewer bytes to store `-1` as `not(0)` than as pure data).
1. **Constraining the design**. We do depend on that in a few key places in the language. For example:
    - Custom operators: custom operator definitions must be `pure`. It would not make sense to require them to be comptime. On the other hand `view` does not restrict them enough - no one expects `+` to depend on blockchain state.
    - Virtual functions: by marking a virtual function `pure` in a reusable base contract you can force its users to uphold that guarantee when they override that function.
1. **Sanity checks**: by declaring a function `pure`, you inform the compiler of the fact that it should not allow you to do certain things in the definition. Or by using a `constant` you tell it that you should not be able to assign to it. Such mistakes should be caught by testing, but just like types or assertions, it's a convenient and lightweight mechanism to express constraints right in your code and catch simple mistakes.
1. **Auditability**: Code is easier to read and understand when it's more "local", i.e. when fewer things can interact with it at a distance. Declaring a function `pure` is a guarantee that lets the reader immediately rule out whole classes of problems without leaving the definition site.

Comptime has all of these benefits, but also has the narrowest range of applicability. The less restrictive levels can in some form still provide all of the above except (1). We definitely need comptime, but IMO it should be a new, distinct level rather than something that replaces `pure`.

> - `pure` precompiles are independent of a chain's state, so it's fine to allow them

They are not. They are a part of the state in the same way as deployed contracts. Different chains can have different sets of precompiles (and not just theoretically - L2s actually do). That's why, for example, we would be fine making `sha256()` comptime, as that's guaranteed by the language to always be the SHA algorithm, but not precompile `0x02` which is not defined in the language and could be anything.

Also, even implementing all the mainnet precompiles in the compiler as language builtins would add a lot of complexity we don't want.

> So, the compiler should use an EVM interpreter (it can be greatly simplified from the "full EVM" interpreter, though).

This is one of the ways to implement comptime, though not the only one. You could just as well interpret Solidity. Interpreting EVM assembly has some advantages (obvious way to handle memory and assembly), but also some downsides (comptime memory potentially leaking to runtime context) and higher implementation complexity. The complexity is why I don't think it's realistic for Classic Solidity (see https://github.com/argotorg/solidity/issues/3157#issuecomment-3293684297), but it's being considered for Core Solidity.
